### ISSUE TITLE
&nbsp;  escaped to " "

### ISSUE THREAD

**Issue by yuyujulin** [OWASP/java-html-sanitizer#87]
For example, I have the following input:
`<input name="test" value="test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob"/><br/>`

After sanitizing, it became the "test        bob", but when we displaying it into a `<p>` tag,
it will be `<p>test bob</p>`, this is not what we want as the count of blank is not 8, can someone look at this issue?

**Comment by mikesamuel** [comment]
I get `<input name="test" value="test        bob" /><br />` and the following testcase passes

```java
  @Test
  public static final void testNbsps() {
    String input =
        "<input name=\"test\""
        + " value=\"test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob\"/>"
        + "<br/>";

    PolicyFactory policy = new HtmlPolicyBuilder()
        .allowElements("input", "br")
        .allowAttributes("name", "value").onElements("input")
        .toFactory();

    String got = policy.sanitize(input);
    System.err.println(got);
    int testEnd = got.indexOf("value=\"test") + "value=\"test".length();
    int bobStart = got.indexOf("bob");
    assertEquals(8, bobStart - testEnd);
  }
```

What makes you believe the count of blanks is not 8?

**Comment by yuyujulin** [comment]
@mikesamuel samuel  thanks for your reply, I mean we just want to save the content as what it was, the content is  "test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob", after sanitizing, i want to save it as "test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob" instead of "test        bob" into DB, could you please provide some workarounds for this?

**Comment by mikesamuel** [comment]
How is html with `&nbsp;` causing trouble in the DB?

**Comment by yuyujulin** [comment]
@mikesamuel  html with `test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob` doesn't cause trouble, but the string `test        bob` causing the trouble, I mean I just want to save the string  `test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob`

**Comment by xenoterracide** [comment]
The reason you would want to do this is so the browser will render as `test        bob` so users would see the very long space...

**Comment by mikesamuel** [comment]
@xenoterracide as demonstrated at https://github.com/OWASP/java-html-sanitizer/issues/87#issuecomment-242238403 the users will see the long space with the `&nbsp;`.

**Comment by xenoterracide** [comment]
```
  @Test
  public static final void testNbsps() {
    String input =
        "<input name=\"test\""
        + " value=\"test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob\"/>"
        + "<br/>";

    PolicyFactory policy = new HtmlPolicyBuilder()
        .allowElements("input", "br")
        .allowAttributes("name", "value").onElements("input")
        .toFactory();

    String got = policy.sanitize(input);
    System.err.println(got);
   
    assertEquals("test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob", got);
  }
```
^ I believe is the desired behavior, and what we would want too.

**Comment by mikesamuel** [comment]
Why?

**Comment by xenoterracide** [comment]
ya, know, that's a good question... I'm looking at it now and I have no idea why I think this is necessary. I'm betting I encountered some reason (or maybe just the same flaw elsewhere)  at some point maybe it had something to do with [this](http://stackoverflow.com/a/6730545/206466)? (wasn't that long ago we were suppoting IE 8, and we have support for 9 ) (speculating IE is usually the culprit, but I don't have access to anything older than 11 here, so I might have to try some stuff tomorrow )

**Comment by johnjaylward** [comment]
This test fails, and I'm not sure why. 
```java
public static final PolicyFactory SANITIZER = Sanitizers.BLOCKS
                .and(Sanitizers.FORMATTING)
                .and(Sanitizers.STYLES)
                .and(Sanitizers.LINKS)
                .and(Sanitizers.TABLES)
                // use custom image sanitation to allow data (base64 inline) images
                // .and(Sanitizers.IMAGES)
                .and(new HtmlPolicyBuilder()
                        .allowUrlProtocols("data", "http", "https")
                        .allowElements("img")
                        .allowAttributes("alt", "src")
                        .onElements("img")
                        .allowAttributes("border", "height", "width")
                        .matching(intAttrPolicy)
                        .onElements("img")

                        .toFactory())
                // allow older font tag even though it's not valid HTML 5
                .and(new HtmlPolicyBuilder()
                        .allowElements("font")
                        .allowAttributes("color", "face")
                        .onElements("font")
                        .allowAttributes("size")
                        .matching(intAttrPolicy)
                        .onElements("font")
                        .toFactory());

    @Test
    public void testStringEquals() {
        final String extected = "\u0160";
        assertTrue(SANITIZER.sanitize("&nbsp;").equals(extected));
    }
```

I see that the sanitizer is not keeping the entity for `&nbsp;`, but it also doesn't appear to be converting it to an expected value.

It would be nice if the sanitizer could just keep the entity instead of replacing it with the unicode value. Is there a way to tell the sanitizer to keep entities?

**Comment by johnjaylward** [comment]
never mind, I had my java unicode wrong. I was using base10 when I should have used base16.

**Comment by mikesamuel** [comment]
@johnjaylward  I'm glad you figured out your problem.  I'm closing this for lack of a motivating use case.

**Comment by fvschie** [comment]
There's a semantic difference between non-breaking spaces and regular spaces. An &nbsp; between two words will prevent line-breaking, which can be important in some applications (email templates, for example, or anywhere where you want to keep two words on the same line for clarity).

Non-breaking spaces should be preserved.

**Comment by xenoterracide** [comment]
I didn't end up using this library... but hindsight being 20/20 JSoup converts &nbsp; to the unicode variant of a non breaking space, and this behavior is superior.

**Comment by jmanico** [comment]
Thanks for that. If you see anything in JSoup that we should be doing,
we're always open to advice.

Aloha, Jim

**Comment by mikesamuel** [comment]
> Non-breaking spaces should be preserved.

@fvschie I agree with everything you said.  When are non-breaking spaces not preserved?

IIUC, the following HTML fragments are equivalent, so switching between them does preserve non-breaking spaces:

* `before&nbsp;after`
* `before&#xA0;after`
* `before&#160;after`
* `before_after` where \_ is U+A0

So when @xenoterracide says

```java
assertEquals("test&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;bob", got);
```

I disagree because I think that `"test\u00A0\u00A0\u00A0\u00A0\u00A0\u00A0\u00A0\u00A0bob"` is a valid output.

**Comment by xenoterracide** [comment]
I would agree with that at this juncture. Jsoup's behavior of converting everything to unicode that it can is preferred by me at this point. The only things it doesn't convert is `&<>` (it encodes those), and I kind of wish there was a unicode variant of those that weren't the ascii ones, so they would look the same but not be interpreted as source code.

**Comment by xenoterracide** [comment]
I'll be honest though, I looked at this library because we had problems with javascript injection, it turned out our problems with actually caused by over sanitization and then cleaning... using esapi (I know this isn't esapi, but was looking at it as a replacement) 'caused us to create more vulnerabilities than would have existed if we had just removed it. Due to improper use and knowledge of how encoding actually needs to work.

**Comment by jmanico** [comment]
Thanks for sharing this.

FYI: ESAPI also uses AntiSamy, make sure you update AntiSamy under the
hood if you're using ESAPI.

- Jim

**Comment by mikesamuel** [comment]
> Jsoup's behavior of converting everything to unicode that it can is preferred by me at this point.

@xenoterracide Where does the sanitizer not do this?

I appreciate the time you've spent explaining your problem, but I have little confidence that I understand what you're seeing or want based on strings with spaces copied into Github markdown.  I don't believe that copy/pasting the string you pasted into the markdown editor gets me the same content, so I don't see that reports like "will render as `test bob` so users" are going to move us towards consensus.

Your testcase asserts that "the count of blanks is not 8" and I gave you a test that shows that is not the case.  Please supply an improved testcase if you believe this is in error.

**Comment by xenoterracide** [comment]
yeah, I've been trying to drop it completely... 

one last thing that might be learned from jsoup, and I don't remember if this does this... but when I run cleans/encodes with jsoup idempotently. meaning if I run the same string through multiple times, it'll result in the same set of encoding.

> Your testcase asserts that "the count of blanks is not 8" and I gave you a test that shows that is not the case. Please supply an improved testcase if you believe this is in error.

that was 2 years ago... IIRC after this, I discovered the double sanitization and normalization 
 problem and so my guess is that what I was getting was actually getting was the result of having my strings run through the esapi string normalizer (which turns `&lt;` into `<`, and then cleaned... Took me a good year of effort to address most of our JS injection problems...

Honestly I'm not sure if I actually ever encountered a real problem at this point (outside of bad code on our end) so, my bad, sorry. This ticket just came across my radar again. Reading this ticket though it's not actually clear to me what the behavior actually is. Your test above doesn't actually show me what characters are present in regards to whitespace, only the count.

**Comment by xenoterracide** [comment]
benefits of unicode. 

* size, `\u00A0` is 2 bytes `&nbsp;` is 6 bytes.
* no double encoding/cleaning issues. (you're unlikely to end up with &amp; on the screen)
* no cut in half issues (oh yeah, I what a string to show 6 characters... so I substring it, later `The&nbs` getting a call from product owner...

**Comment by xenoterracide** [comment]
> but that may be because of the unholy hybrid of JSoup, Java-html-sanitizer and I believe Tidy 
sounds likely... I would check each transition to see what the input and outputs are... find out exactly which method starts with multiple `\u00a0` and ends with one.

**Comment by mikesamuel** [comment]
@fvschie I edited your comment to use code formatting for entities.

> For HTML validation, what's the benefit of replacing it with unicode?

This tool doesn't do HTML validation.

I could have chosen to encode all non-7bit ASCII characters, but did not because that introduces a lot of overhead for [CJK](https://en.wikipedia.org/wiki/CJK_characters) text.

As you've pointed out, there are many non-printable, but significant characters.  Picking and choosing which to encode seems hard, so once I decided not to encode all, it was easy to decide to encode a minimal set of codepoints.

The rule that decide what to encode are:

<details><summary>ASCII Characters encoded</summary>

https://github.com/OWASP/java-html-sanitizer/blob/f07e44b034a45d94d6fd010279073c38b6933072/src/main/java/org/owasp/html/Encoding.java#L319-L354

</details>

<details><summary>Gnarly rules for non-ASCII code units</summary>

https://github.com/OWASP/java-html-sanitizer/blob/f07e44b034a45d94d6fd010279073c38b6933072/src/main/java/org/owasp/html/Encoding.java#L206-L283

</details>

**Comment by jmanico** [comment]
Backlog triage ahead of the next release.

Answered thoroughly in-thread. `&nbsp;` is decoded to U+00A0 (the actual non-breaking space character), not to a plain space — the two are semantically identical in HTML, and the non-breaking behaviour is preserved. `before&nbsp;after`, `before&#xA0;after` and `before<U+00A0>after` are the same document.

Emitting the named entity back out is a rendering preference rather than a sanitization concern, and the library deliberately does not entity-encode non-ASCII (it would bloat CJK text considerably). Thanks @mikesamuel, @fvschie and @xenoterracide for working through it.
